src/punycode.cpp

```diff
@@ -0,0 +1,338 @@
+#include <algorithm>
+#include <string>
+#include <iostream>
+
+#include "punycode.h"
+#include "utf8.h"
+
+namespace Url
+{
+
+    const std::vector<int8_t> Punycoder::BASIC_TO_DIGIT =
+    {
+        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+
+        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+        26, 27, 28, 29, 30, 31, 32, 33, 34, 35, -1, -1, -1, -1, -1, -1,
+
+        -1,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14,
+        15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, -1, -1, -1, -1, -1,
+
+        -1,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14,
+        15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, -1, -1, -1, -1, -1,
+
+        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+
+        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+
+        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+
+        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1
+    };
+
+    const std::string Punycoder::DIGIT_TO_BASIC = "abcdefghijklmnopqrstuvwxyz0123456789";
+
+    std::string& Punycoder::encode(std::string& str)
+    {
+        // Pseudocode copied from https://tools.ietf.org/html/rfc3492#section-6.3
+        //
+        // let n = initial_n
+        // let delta = 0
+        // let bias = initial_bias
+        Utf8::codepoint_t n = INITIAL_N;
+        long delta = 0;
+        long bias = INITIAL_BIAS;
+        std::string output;
+
+        // Accumulate the non-basic codepoints
+        std::vector<Utf8::codepoint_t> codepoints;
+        for (auto it = str.cbegin(); it != str.cend(); )
+        {
+            long value = Utf8::readCodepoint(it, str.cend());
+            if (value < 0x80)
+            {
+                // copy them to the output in order
+                output.append(1, static_cast<char>(value));
+            }
+            codepoints.push_back(value);
+        }
+
+        // let h = b = the number of basic code points in the input
+        size_t h = output.size();
+        size_t b = h;
+
+        // copy a delimiter if b > 0
+        if (b > 0)
+        {
+            output.append(1, '-');
+        }
+    
+        // while h < length(input) do begin
+        while (h < codepoints.size())
+        {
+            // let m = the minimum {non-basic} code point >= n in the input
+            long m = std::numeric_limits<long>::max();
+            for (auto it = codepoints.begin(); it != codepoints.end(); ++it)
```

Comment:
Use [`std::max_element`](http://en.cppreference.com/w/cpp/algorithm/max_element). You can compare against `n` after the fact.